communicator/room.go

```diff
@@ -206,64 +245,125 @@ func (r *Room) startStorePhase() {
 }
 
 func (r *Room) startBattlePhase() {
-	var clients []*Client
-	for client := range r.playersState {
-		clients = append(clients, client)
+	type PlayerPair struct {
+		players  []*Client
+		loserIdx int
+		finished bool
 	}
+	var alreadyMatched []*Client
+	var playerPairs []*PlayerPair
+	for client := range r.playersState {
+		playerPair := &PlayerPair{
+			players: []*Client{client, r.getEnemy(client)},
+		}
+		if clientInSlice(client, alreadyMatched) || playerPair.players[1] == nil {
+			continue
+		}
 
-	player1State := r.playersState[clients[0]]
-	player1 := PlayerBattleSetup{
-		UnitPlacement: player1State.UnitsPlacement,
-		Units:         player1State.Units,
+		alreadyMatched = append(alreadyMatched, playerPair.players[0])
+		alreadyMatched = append(alreadyMatched, playerPair.players[1])
+		playerPairs = append(playerPairs, playerPair)
 	}
 
-	player2State := r.playersState[clients[1]]
-	player2 := PlayerBattleSetup{
-		UnitPlacement: mirrorUnitPlacements(player2State.UnitsPlacement),
-		Units:         player2State.Units,
-	}
+	for _, playerPair := range playerPairs {
+		player1State := r.playersState[playerPair.players[0]]
+		player1 := PlayerBattleSetup{
+			UnitPlacement: player1State.UnitsPlacement,
+			Units:         player1State.Units,
+		}
 
-	battleResult, err := GetBattleResult(player1, player2)
-	if err != nil {
-		r.Shutdown("Failed fetching battle result")
-		r.log.Println("Failed fetching battle result", err)
-		return
-	}
+		player2State := r.playersState[playerPair.players[1]]
+		player2 := PlayerBattleSetup{
+			UnitPlacement: mirrorUnitPlacements(player2State.UnitsPlacement),
+			Units:         player2State.Units,
+		}
 
-	var finished bool
-	for i, c := range clients {
-		if i != battleResult.Winner {
-			r.playersState[c].Player.Hp += battleResult.PlayerHpChange
+		battleResult, err := GetBattleResult(player1, player2)
+		if err != nil {
+			r.Shutdown("Failed fetching battle result")
+			r.log.Println("Failed fetching battle result", err)
+			return
+		}
 
-			if r.playersState[c].Player.Hp <= 0 {
-				finished = true
+		for i, c := range playerPair.players {
+			if i != battleResult.Winner {
+				playerPair.loserIdx = i
+				r.playersState[c].Player.Hp += battleResult.PlayerHpChange
+				if r.playersState[c].Player.Hp <= 0 {
+					playerPair.finished = true
+				}
 			}
 		}
-	}
 
-	for i, c := range clients {
-		battleStatistics := BattleStatistics{
-			Log: battleResult.Log,
+		for i, c := range playerPair.players {
+			battleStatistics := BattleStatistics{
+				Log: battleResult.Log,
+			}
+			if i == battleResult.Winner {
+				battleStatistics.Result = GameResultWin
+			} else {
+				battleStatistics.Result = GameResultLoss
+				battleStatistics.PlayerHpChange = battleResult.PlayerHpChange
+			}
+
+			if err := c.SendMessage(r.generateClientState(c, nil, &battleStatistics)); err != nil {
+				r.Shutdown("Failed to propagate client state")
```

Comment:
I don't think we should shutdown the room if there is more than 1 player remaining. The client should be disconnected and the game should continue.